frontend/src/util/apiWrapper.js

```diff
@@ -0,0 +1,178 @@
+import { Semaphore } from 'await-semaphore';
+import axios from 'axios';
+import { createDefaultBundleName, pathIsArchive, getArchiveExt } from './worksheet_utils';
+
+export const get = (url, params) => {
+    const requestOptions = {
+        params,
+    };
+    return axios.get(url, requestOptions).then((res) => res.data);
+};
+
+export const post = (url, data, config) => {
+    return axios.post(url, data, config).then((res) => res.data);
+};
+
+export const put = (url, data) => {
+    return axios.put(url, data).then((res) => res.data);
+};
+
+export const patch = (url, data) => {
+    return axios.patch(url, data).then((res) => res.data);
+};
+
+// prefixed with underscored because delete is a reserved word in javascript
+export const _delete = (url) => {
+    return axios.delete(url).then((res) => res.data);
+};
+
+export const updateEditableField = (url, data) => {
+    return patch(url, data);
+};
+
+export const getUser = () => {
+    return get('/rest/user');
+};
+
+export const getUsers = (username) => {
+    const url = '/rest/users/' + username;
+    return get(url);
+};
+
+export const updateUser = (data) => {
+    const url = '/rest/user';
+    return patch(url, { data });
+};
+
+export const navBarSearch = (keywords) => {
+    const url = '/rest/interpret/wsearch';
+    return post(url, { keywords });
+};
+
+export const addItems = (worksheetUUID, data) => {
+    const url = `/rest/worksheets/${worksheetUUID}/add-items`;
+    return post(url, data);
+};
+
+export const executeCommand = (command, worksheet_uuid) => {
+    // returns a Promise
+    const url = '/rest/cli/command';
+    return post(url, {
+        worksheet_uuid: worksheet_uuid || null,
+        command: command,
+    });
+};
+
+export const completeCommand = (command, worksheet_uuid) => {
+    // returns a Promise
+    const url = '/rest/cli/command';
+    return post(url, {
+        worksheet_uuid: worksheet_uuid || null,
+        command: command,
+        autocomplete: true,
+    });
+};
+
+// Limit concurrent requests for async resolving items
+const MAX_CONCURRENT_REQUESTS = 3;
+const semaphore = new Semaphore(MAX_CONCURRENT_REQUESTS);
+
+export const fetchAsyncBundleContents = async ({ contents }) => {
+    // used in table and record items
+    return semaphore.use(async () => {
```

Comment:
probably my naivete but why do we want to use this as opposed to https://www.npmjs.com/package/await-semaphore#promise-style-javascript

Might be more consistent with the rest of the codebase, maybe not.

Reply:
This automatically handles acquire/release, the behavior should be the same as other styles.